src/repl.ml

```diff
@@ -192,11 +192,20 @@ end
 
 let register extension instance =
   let disposable =
-    Vscode.Window.onDidCloseTerminal
-      ()
-      ~listener:(fun terminal ->
-        if String.equal (Vscode.Terminal.name terminal) name then
-          Extension_instance.close_repl instance)
-      ()
+    let listener closed_terminal =
+      ignore
+        (match Extension_instance.repl instance with
+        | None -> Promise.return ()
+        | Some repl_terminal ->
+          let open Promise.Syntax in
+          let+ repl_terminal_pid, closed_terminal_pid =
+            Promise.all2
+              ( Terminal.processId repl_terminal
+              , Terminal.processId closed_terminal )
+          in
+          if repl_terminal_pid = closed_terminal_pid then
```

Comment:
In general, relying on pid's isn't the most reliable thing because they can always be reused by the OS. So I think that adding our own unique ID would still be helpful. I don't think this is essential though.

Reply:
> In general, relying on pid's isn't the most reliable thing because they can always be reused by the OS.

But this PID would be provided by vscode, so it would require vscode being buggy?

> So I think that adding our own unique ID would still be helpful. 

I didn't quite understand your suggestion tbh because `closed_terminal` has type `Terminal.t` which is a vscode type, which we can't add an ID to?

Reply:
> But this PID would be provided by vscode, so it would require vscode being buggy?

PID's are ultimately provided by the the OS. So in theory it's possible that our listener can get called with a terminal that has the same PID but isn't our repl. This can happen if our terminal is closed ,and a new one is immediately opened and closed, and the OS assigned it the same PID. I don't know the vscode api well enough to know if this is indeed a possibility, but it is a reason why relying on dead PIDs is considered fragile.

> I didn't quite understand your suggestion tbh because closed_terminal has type Terminal.t which is a vscode type, which we can't add an ID to?

Yes, that's my mistake. I thought we were dealing with `Terminal_sandbox.t` here, but this is just a type alias to the vscode type.

Reply:
I think you're right, and relying on PID is a mess because `processId` _can_ return an undefined based on `vscode.d.ts` rather than online docs. 

I had a look at the terminal service implementation in vscode's extension host, and from the looks, terminal objects should be physically equal. Should I go back to it? Why do you think physical eq is bad given most of the js development is based on `===` which is (?) equivalent to physical eq for objects?